### PR TITLE
New fields and flexible label-column matching with regexps.

### PR DESCRIPTION
**Pull request by alexishuf**
Programmable web made slight changes to labels shown in API description that break the current field assignment. To avoid constant updates on the fields array, regexps are generated from that array in order to match the labels used on the scrapped HTML.

New fields used on descriptions were added as well, no field was renamed in the database so queries written for the database in previous version should still be compatible. The only breaking change on ProgrammableWeb side was 'API Homepage' changed to 'Api Portal Home Page'. Together with the aforementioned regexps, a 'Home_Page' -> 'Homepage' rule was added to keep things working.

Ran the scrapper locally on 2016-08-10. The only errors were 16 redirects.

### CODE REVIEW

**Comment by IvanGoncharov** [comment]
@alexishuf Great, need some time to understand all changes 😄

File: scraper.js [hunk 1]
```diff
@@ -94,6 +125,33 @@ function getFollowers($) {
...
 }
 
+function preprocessLabel(label) {
+  var old = null;
+  while (old !== label) {
+    old = label;
+    _([ ['/', '_'], ['-', '_'], ['+', '_'], [/ +/, '_'], [/__+/, '_'],
+        ['Home_Page', 'Homepage'],
+        [/[ _](is|this|the|an?|for|in|on)[ _]/i, ' '],
+        [/^(is|this|the|an?|for|in|on)[ _]/i, ' '],
```
Comment (IvanGoncharov) [inline scraper.js]: You can merge two regex:`/(^|[ _])...`

File: scraper.js [hunk 2]
```diff
@@ -94,6 +125,33 @@ function getFollowers($) {
...
+        [/[.;,?]+/, ''], [/[ _]$/, ''], [/^[ _]/, ''] ])
+      .each(function (args) {
+        var tmp = label;
+        label = label.replace(args[0], args[1]);
+      });
+  }
+  return label;
+}
+
+
```
Comment (IvanGoncharov) [inline scraper.js]: Two lines

File: scraper.js [hunk 3]
```diff
@@ -94,6 +125,33 @@ function getFollowers($) {
...
 
+function preprocessLabel(label) {
+  var old = null;
+  while (old !== label) {
+    old = label;
+    _([ ['/', '_'], ['-', '_'], ['+', '_'], [/ +/, '_'], [/__+/, '_'],
+        ['Home_Page', 'Homepage'],
+        [/[ _](is|this|the|an?|for|in|on)[ _]/i, ' '],
+        [/^(is|this|the|an?|for|in|on)[ _]/i, ' '],
+        [/[.;,?]+/, ''], [/[ _]$/, ''], [/^[ _]/, ''] ])
```
Comment (IvanGoncharov) [inline scraper.js]: It better to use `strip` for whitespaces

**Comment by IvanGoncharov** [comment]
@alexishuf I really appriciate effort that you in this PR.
But the problem that you trying to solve is matching 20-30 strings to column names.
For that you wrote really complicated code it took me ~30 min to understand what's going on.
Moreover, there is no guarantee that it will work if Programmable web will do even minor changes like uppercase  to lowercase conversion.

If the programmable web had hundreds of fields it may be a valuable solution.
But for small number of strings to be mapped it way easier to write simple map like:
```js
var dynamicFields = {
 Twitter_Url: 'Twitter URL',
 ...
}
```
It's way more simple solution and easier to maintain.

**Comment by alexishuf** [comment]
@IvanGoncharov Thanks for the comments.

Many changes on field names from the last successful run on morph.io were really dummy, that is why I opted for the regexp approach. Regexps would hopefully avoid the scrapper breaking from minor changes on the HTML labels. The only downside would be eventually wrong mappings (which happened on my first run), that's the reason for that weird select largest matched field.

But it is more a matter of the project style and goals. I can do it as a map later today and submit a new PR, unless you want to do it yourself.

**Comment by IvanGoncharov** [comment]
> But it is more a matter of the project style and goals.

I wrote it for my own experiment and though I can use ProgrammableWeb data for my main project:
https://github.com/APIs-guru/api-models
But I quickly found out that ProgrammableWeb doesn't keep data up-to-date.
I personally don't use it anymore but it could be useful for others.
So I like it to be as maintainable as easy as possible.

> I can do it as a map later today and submit a new PR

Yes, it would be great.
If needed you can use `trim`.

**Comment by alexishuf** [comment]
PR #2 solves the problem with a map.

**Pull request closed without merging**